IOperation and CFG support for null-coalescing assignment

333fred opened the pull request:
Added IOperation and CFG support and tests for null-coalescing assignments. Also added tests for semantic model APIs. @dotnet/roslyn-compiler @cston @dotnet/analyzer-ioperation for review.

cston commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -1292,6 +1294,19 @@ private ICoalesceOperation CreateBoundNullCoalescingOperatorOperation(BoundNullC
             return new LazyCoalesceExpression(expression, whenNull, valueConversion, _semanticModel, syntax, type, constantValue, isImplicit);
         }
 
+        private IOperation CreateBoundNullCoalescingAssignmentOperatorOperation(BoundNullCoalescingAssignmentOperator boundNode)
```
>OperatorOperation [](start = 62, length = 17)

`OperatorOperation`?

333fred replied:
Yep. The format in this file is `Create` + `BoundNodeName` + `Operation`.

---
In reply to: [207689657](https://github.com/dotnet/roslyn/pull/29063#discussion_r207689657) [](ancestors = 207689657)

cston commented on src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs:
```diff
@@ -3790,6 +3790,71 @@ internal sealed partial class LazyCoalesceExpression : BaseCoalesceExpression, I
         public override IOperation WhenNull => SetParentOperation(_lazyWhenNull.Value, this);
     }
 
+    internal abstract partial class BaseCoaleseAssignmentOperation : Operation, ICoalesceAssignmentOperation
```
>Coalese [](start = 40, length = 7)

Typo

cston commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ICoalesceAssignmentOperation.cs:
```diff
@@ -0,0 +1,895 @@
+        }
+
+        [Fact]
+        public void CoalesceAssignmentFlow_WhenNullConversion()
```
>CoalesceAssignmentFlow_WhenNullConversion [](start = 20, length = 41)

Is this different in terms of a flow graph from `CoalesceAssignmentFlow_SimpleAssignment`? If not, consider removing one of the two tests.

333fred replied:
There is an implicit conversion in the WhenNull side of this one. I can remove the SimpleAssignment version, as it doesn't exercise anything this does not.

---
In reply to: [207691151](https://github.com/dotnet/roslyn/pull/29063#discussion_r207691151) [](ancestors = 207691151)

jaredpar commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ICoalesceAssignmentOperation.cs:
```diff
@@ -0,0 +1,895 @@
+            Statements (1)
+                IFlowCaptureOperation: 1 (OperationKind.FlowCapture, Type: null, IsImplicit) (Syntax: 'o1')
+                  Value: 
+                    IFlowCaptureReferenceOperation: 0 (OperationKind.FlowCaptureReference, Type: System.Object, IsImplicit) (Syntax: 'o1')
```
Why is there a flow capture of a flow capture here?

333fred replied:
Flow capture id 0 is just for the IsNull check. Flow capture 1 is for the evaluation of the expression.

---
In reply to: [207938024](https://github.com/dotnet/roslyn/pull/29063#discussion_r207938024) [](ancestors = 207938024)

333fred replied:
@AlekseyTs for an opinion here. This is an optimization we _could_ make, but I don't think we have another scenario in the CFG where we capture a result, do a test on the capture, and then re-capture in the false case.

---
In reply to: [207965573](https://github.com/dotnet/roslyn/pull/29063#discussion_r207965573) [](ancestors = 207965573,207938024)

AlekseyTs replied:
> This is an optimization we could make, but I don't think we have another scenario in the CFG where we capture a result, do a test on the capture, and then re-capture in the false case.

Yes, we intentionally avoid assigning to the same capture more than once on the same code path in the graph.

---
In reply to: [209018400](https://github.com/dotnet/roslyn/pull/29063#discussion_r209018400) [](ancestors = 209018400,207965573,207938024)

jcouv commented:
Waiting to hear back from IOperation design team. 
@333fred May be good to ping the thread. Thanks

mavasani commented:
> Waiting to hear back from IOperation design team.

I believe this was already discussed at last week's meeting and Fred's implementation matches the agreed design.

jcouv commented:
@mavasani Sorry, I commented on the wrong PR :-S

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ICoalesceAssignmentOperation.cs:
```diff
@@ -0,0 +1,34 @@
+    /// This interface is reserved for implementation by its associated APIs. We reserve the right to
+    /// change it in the future.
+    /// </remarks>
+    public interface ICoalesceAssignmentOperation : IOperation
```
>IOperation [](start = 52, length = 10)

Should we inherit from IAssignmentOperation instead and use ```Value``` from it?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ICoalesceAssignmentOperation.cs:
```diff
@@ -0,0 +1,34 @@
+        /// <summary>
+        /// Whether the assignment is evaluated as a checked operation. This only has meaning in dynamic contexts.
+        /// </summary>
+        bool IsChecked { get; }
```
>IsChecked [](start = 13, length = 9)

Could you provide more details how an assignment can be checked or unchecked?

333fred replied:
After investigation, I will remove this field altogether.

---
In reply to: [209380732](https://github.com/dotnet/roslyn/pull/29063#discussion_r209380732) [](ancestors = 209380732)

AlekseyTs commented on src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs:
```diff
@@ -3790,6 +3790,71 @@ internal sealed partial class LazyCoalesceExpression : BaseCoalesceExpression, I
+        {
+            get
+            {
+                if (Target != null)
```
>Target [](start = 20, length = 6)

Can ```Target``` ever be null?

333fred replied:
Doesn't look like it. I'll add asserts to the verifier, but I'm going to keep the `if` checks as we do that with every other field in this file.

---
In reply to: [209382051](https://github.com/dotnet/roslyn/pull/29063#discussion_r209382051) [](ancestors = 209382051)

333fred replied:
I'll also add a test for otherwise empty left and right nodes.

---
In reply to: [210781658](https://github.com/dotnet/roslyn/pull/29063#discussion_r210781658) [](ancestors = 210781658,209382051)

AlekseyTs commented on src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs:
```diff
@@ -3790,6 +3790,71 @@ internal sealed partial class LazyCoalesceExpression : BaseCoalesceExpression, I
+                {
+                    yield return Target;
+                }
+                if (WhenNull != null)
```
>WhenNull [](start = 20, length = 8)

Can ```WhenNull``` ever be null?

333fred replied:
Same as above.

---
In reply to: [209382124](https://github.com/dotnet/roslyn/pull/29063#discussion_r209382124) [](ancestors = 209382124)

AlekseyTs commented on src/Test/Utilities/Portable/Compilation/ControlFlowGraphVerifier.cs:
```diff
@@ -783,10 +783,14 @@ PooledHashSet<CaptureId> adjustAndGetEntryState(ArrayBuilder<PooledHashSet<Captu
                         $"Operation [{operationIndex}] in [{getBlockId(block)}] uses not initialized capture [{id.Value}].");
 
+                    // Except for a few specific scenarios, any references to captures should either be long-lived capture references,
+                    // or they should come from the enclosing region. Otherwise, it signifies that the region packing algorithm failed
```
>it signifies that the region packing algorithm failed [](start = 81, length = 53)

If that refers to the pack pass at the end of the graph building process, I think this is not accurate. If I remember correctly, it is not responsible for enforcing "proper" distribution of captures among regions.

AlekseyTs commented on src/Test/Utilities/Portable/Compilation/ControlFlowGraphVerifier.cs:
```diff
@@ -825,6 +829,18 @@ bool isConditionalAccessReceiver(IFlowCaptureReferenceOperation reference)
+            bool isCoalesceAssignmentTarget(IFlowCaptureReferenceOperation reference)
+            {
+                SyntaxNode referenceSyntax = reference.Syntax;
+                if (referenceSyntax.Parent.IsKind(CSharp.SyntaxKind.ParenthesizedExpression))
```
>if (referenceSyntax.Parent.IsKind(CSharp.SyntaxKind.ParenthesizedExpression)) [](start = 16, length = 77)

Is it sufficient to check just one level? Should we use ```applyParenthesizedIfAnyCS``` helper instead?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2717,12 +2718,11 @@ private static void LinkBlocks(BasicBlockBuilder previous, IOperation condition,
         /// Caller is responsible for spilling the stack and pushing a stack frame before calling this helper.
         /// </summary>
-        private IOperation NullCheckAndConvertCoalesceValue(ICoalesceOperation operation, BasicBlockBuilder whenNull)
+        private IOperation NullCheckAndConvertCoalesceValue(IOperation operationValue, IConvertibleConversion coalesceConversion, ITypeSymbol coalesceType, BasicBlockBuilder whenNull)
```
I am also not sure if it makes sense to reuse this function for CoalesceAssignment at all. It might be better to extract its first part (that deals with branching) into a separate helper and avoid dealing with faking conversions 

---
In reply to: [209388084](https://github.com/dotnet/roslyn/pull/29063#discussion_r209388084) [](ancestors = 209388084)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2835,6 +2837,53 @@ public override IOperation VisitCoalesce(ICoalesceOperation operation, int? capt
+            IOperation testExpression = NullCheckAndConvertCoalesceValue(operation.Target, coalesceConversion: null, operation.Type, whenNull);
+
+            // Coalesce can have conversions on the left side, but coalesce assignment should not.
+            Debug.Assert(testExpression.Kind != OperationKind.Conversion);
```
>Debug.Assert(testExpression.Kind != OperationKind.Conversion); [](start = 12, length = 62)

Can ```testExpression``` be a conversion in  an error case?

333fred replied:
No longer relevant, as I'm not using `NullCheckAndConvertCoalesceValue` anymore.

---
In reply to: [209390571](https://github.com/dotnet/roslyn/pull/29063#discussion_r209390571) [](ancestors = 209390571)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2835,6 +2837,53 @@ public override IOperation VisitCoalesce(ICoalesceOperation operation, int? capt
+
+            int resultCaptureId = captureIdForResult ?? GetNextCaptureId(resultCaptureRegion);
+
+            AddStatement(new FlowCapture(resultCaptureId, operation.Target.Syntax, testExpression));
```
>operation.Target.Syntax [](start = 58, length = 23)

```operation.Syntax``` ?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2835,6 +2837,53 @@ public override IOperation VisitCoalesce(ICoalesceOperation operation, int? capt
             return result;
         }
 
+        public override IOperation VisitCoalesceAssignment(ICoalesceAssignmentOperation operation, int? captureIdForResult)
```
>VisitCoalesceAssignment [](start = 35, length = 23)

It doesn't look like ```IsChecked``` used anywhere.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2835,6 +2837,53 @@ public override IOperation VisitCoalesce(ICoalesceOperation operation, int? capt
+            // The return of Visit(operation.WhenNull) can be a flow capture that wasn't used in the non-null branch. We want to create a
+            // region around it to ensure that the scope of the flow capture is as narrow as possible. If there was no flow capture, region
+            // packing will take care of removing the empty region.
+            var whenNullRegion = new RegionBuilder(ControlFlowRegionKind.LocalLifetime, isStackSpillRegion: true);
```
>var whenNullRegion = new RegionBuilder(ControlFlowRegionKind.LocalLifetime, isStackSpillRegion: true); [](start = 12, length = 102)

Does this really help? It feels like ```PopStackFrame(frame);``` below can still merge this region into its parent. It might be simpler and more robust to allocate a new frame and use PopStackFrameAndLeaveRegion for it after ```AddStatement(new FlowCapture(resultCaptureId, operation.Syntax, nullAssignment));```

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ICoalesceAssignmentOperation.cs:
```diff
@@ -0,0 +1,895 @@
+}
+";
+            string expectedOperationTree = @"
+ICoalesceAssignmentOperation (OperationKind.CoalesceAssignment, Type: ?, IsInvalid) (Syntax: 'c1 ??= s1')
```
>Type: ? [](start = 64, length = 7)

I think we can safely say that the type is ```C```

333fred replied:
This follows the same convention that other compound assignments do, where if there is no conversion from the left to the right, the whole compound expression has the error type. See `ICompoundAssignment_BinaryOperatorOutConversion_InvalidMissingInConversion`.

---
In reply to: [209401968](https://github.com/dotnet/roslyn/pull/29063#discussion_r209401968) [](ancestors = 209401968)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ICoalesceAssignmentOperation.cs:
```diff
@@ -0,0 +1,895 @@
+}
+";
+            string expectedOperationTree = @"
+ICoalesceAssignmentOperation(IsChecked: True) (OperationKind.CoalesceAssignment, Type: dynamic) (Syntax: 'd1 ??= d2')
```
>IsChecked: True [](start = 29, length = 15)

It would be good to have explicit tests for unchecked dynamic case and both flavors for non-dynamic.

333fred replied:
No longer relevant, as `IsChecked` has been removed.

---
In reply to: [209402877](https://github.com/dotnet/roslyn/pull/29063#discussion_r209402877) [](ancestors = 209402877)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ICoalesceAssignmentOperation.cs:
```diff
@@ -0,0 +1,895 @@
+        Statements (1)
+            IExpressionStatementOperation (OperationKind.ExpressionStatement, Type: null) (Syntax: 'o1 ??= o2;')
+              Expression: 
+                IFlowCaptureReferenceOperation: 1 (OperationKind.FlowCaptureReference, Type: System.Object, IsImplicit) (Syntax: 'o1 ??= o2')
```
>IFlowCaptureReferenceOperation: 1 (OperationKind.FlowCaptureReference, Type: System.Object, IsImplicit) (Syntax: 'o1 ??= o2') [](start = 16, length = 125)

Should we get rid of this node, its parent and the result capture as well?

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [412596511](https://github.com/dotnet/roslyn/pull/29063#issuecomment-412596511) [](ancestors = 412596511)

333fred commented:
@cston for rereview, as some larger changes have been made.

cston commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2835,6 +2835,76 @@ public override IOperation VisitCoalesce(ICoalesceOperation operation, int? capt
+
+            EvalStackFrame frame = PushStackFrame();
+
+            IOperation testExpression;
```
>testExpression [](start = 23, length = 14)

Consider moving declaration down to assignment.

cston commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2835,6 +2835,76 @@ public override IOperation VisitCoalesce(ICoalesceOperation operation, int? capt
+            EvalStackFrame whenNullFrame = PushStackFrame();
+
+            IOperation whenNullValue = Visit(operation.Value);
+            IOperation nullAssignment = new SimpleAssignmentExpression(OperationCloner.CloneOperation(testExpression), isRef: false, whenNullValue, semanticModel: null,
```
>nullAssignment [](start = 23, length = 14)

The name sounds like we're assigning `null`. Consider renaming. Perhaps `assignment` or `whenNullAssignment`.

333fred commented:
Ping @AlekseyTs

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2835,6 +2835,76 @@ public override IOperation VisitCoalesce(ICoalesceOperation operation, int? capt
+                _currentBasicBlock = null;
+
+                Debug.Assert(testExpression.Kind == OperationKind.FlowCaptureReference);
+                AddStatement(new FlowCapture(resultCaptureId, operation.Syntax, OperationCloner.CloneOperation(testExpression)));
```
>OperationCloner.CloneOperation(testExpression)) [](start = 80, length = 47)

I think this is going to be interpreted as another read from the Target, is this going to accurately reflect what is happening?

333fred replied:
We spill the eval stack, so this is always a capture of a capture. I don't think that that implies another read, since it's a captured value.

---
In reply to: [211441273](https://github.com/dotnet/roslyn/pull/29063#discussion_r211441273) [](ancestors = 211441273)

AlekseyTs replied:
> We spill the eval stack, so this is always a capture of a capture. I don't think that that implies another read, since it's a captured value.

Is this really a value though? How can we assign to it below? Do we have similar pattern in other situations?

---
In reply to: [211450764](https://github.com/dotnet/roslyn/pull/29063#discussion_r211450764) [](ancestors = 211450764,211441273)

333fred replied:
After offline discussion, I'm going to take the approach of separate captures for the "location" and "value" of the left hand side. This is the same approach that the local rewriter takes, and while we don't have the concept of location vs value in captures today, this is something we want to add and this will be compatible when we do.

---
In reply to: [211684153](https://github.com/dotnet/roslyn/pull/29063#discussion_r211684153) [](ancestors = 211684153,211450764,211441273)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2835,6 +2835,76 @@ public override IOperation VisitCoalesce(ICoalesceOperation operation, int? capt
+            LeaveRegionsUpTo(resultCaptureRegion);
+            AppendNewBlock(afterCoalesce);
+
+            return operation.Parent.Kind == OperationKind.ExpressionStatement ? null : GetCaptureReference(resultCaptureId, operation);
```
>operation.Parent.Kind == OperationKind.ExpressionStatement [](start = 19, length = 58)

isStatement?

AlekseyTs commented on src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs:
```diff
@@ -702,6 +702,14 @@ public override void VisitCoalesce(ICoalesceOperation operation)
+        {
+            Assert.Equal(OperationKind.CoalesceAssignment, operation.Kind);
+            Assert.NotNull(operation.Target);
+            Assert.NotNull(operation.Value);
```
These are probably redundant. Elsewhere we check that all items in ```operation.Children``` are not null. In combination with that, the next assert ensures Target and Value are not null.

AlekseyTs commented:
Done with review pass (iteration 5)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [414764088](https://github.com/dotnet/roslyn/pull/29063#issuecomment-414764088) [](ancestors = 414764088)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2835,6 +2835,83 @@ public override IOperation VisitCoalesce(ICoalesceOperation operation, int? capt
+            SpillEvalStack();
+            int valueCaptureId = GetNextCaptureId(valueFrame.RegionBuilderOpt);
+            AddStatement(new FlowCapture(valueCaptureId, locationCapture.Syntax, locationCapture));
+            IOperation valueCapture = GetCaptureReference(valueCaptureId, OperationCloner.CloneOperation(locationCapture));
```
>OperationCloner.CloneOperation [](start = 74, length = 30)

I think clone is not necessary, ```GetCaptureReference``` doesn't incorporate this node into its result.

333fred replied:
I'll address this in a follow-up PR

---
In reply to: [212137397](https://github.com/dotnet/roslyn/pull/29063#discussion_r212137397) [](ancestors = 212137397)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2835,6 +2835,83 @@ public override IOperation VisitCoalesce(ICoalesceOperation operation, int? capt
+            SpillEvalStack();
+            int valueCaptureId = GetNextCaptureId(valueFrame.RegionBuilderOpt);
+            AddStatement(new FlowCapture(valueCaptureId, locationCapture.Syntax, locationCapture));
+            IOperation valueCapture = GetCaptureReference(valueCaptureId, OperationCloner.CloneOperation(locationCapture));
```
>IOperation valueCapture [](start = 12, length = 23)

Would it make sense to introduce ```valueCapture``` only for non-statement case? Only then we are using the value twice.

333fred replied:
I thought about that, but we still need a capture for the location for the when null case, and then we're back in the same problem where locationCapture is being used as both a value and a location in different sections.

---
In reply to: [212138723](https://github.com/dotnet/roslyn/pull/29063#discussion_r212138723) [](ancestors = 212138723)

333fred commented:
@dotnet-bot retest windows_release_unit64_prtest please

333fred commented:
Test failure was a transient file locking issue, if it happens again will file a bug.
